network/parallelism.go

```diff
@@ -0,0 +1,263 @@
+		if time.Since(err) <= p.cfg.ResetInterval {
+			keepSuccesses = append(keepSuccesses, err)
+		}
+	}
+	p.networkSuccesses = keepSuccesses
+	errorRate := p.networkErrorRate()
+	// If we have network errors then ramp down the number of loops.
+	if p.cfg.AllowedNetworkErrorPercent != 0.0 && errorRate >= p.cfg.AllowedNetworkErrorPercent {
+		// Need to keep the value between min and max.
+		if p.currentDesired-1 >= p.cfg.MinConnections {
+			level.Debug(p.l).Log("msg", "triggering lower desired due to network errors", "desired", p.currentDesired-1)
+			p.changeParallelism(p.currentDesired - 1)
+		}
+		return
+	}
+	// If we are drifting too much then ramp up the number of loops.
+	if p.timestampDriftSeconds > p.cfg.AllowedDriftSeconds {
+		// Need to keep the value between min and max.
+		if p.currentDesired+1 <= p.cfg.MaxConnections {
+			level.Debug(p.l).Log("msg", "increasing desired due to timestamp drift", "desired", p.currentDesired+1, "drift", p.timestampDriftSeconds)
+			p.changeParallelism(p.currentDesired + 1)
+		}
+		return
+	}
+
+	// Can we ramp down, if we have dropped below our minimum scale down.
+	if p.timestampDriftSeconds < p.cfg.MinimumScaleDownDriftSeconds {
+		// Need to keep the value between min and max.
+		if p.currentDesired-1 >= p.cfg.MinConnections {
+			level.Debug(p.l).Log("msg", "decreasing desired due to drift lowering", "desired", p.currentDesired-1, "drift", p.timestampDriftSeconds)
+			p.changeParallelism(p.currentDesired - 1)
+		}
+	}
+	level.Debug(p.l).Log("msg", "no changes needed", "desired", p.currentDesired)
+}
+
+func (p *parallelism) networkErrorRate() float64 {
+	// If nothing has happened assume success
+	if len(p.networkSuccesses) == 0 && len(p.networkErrors) == 0 {
+		return 0.0
+	}
+
+	if len(p.networkErrors) == 0 {
+		return 0.0
+	}
+
+	if len(p.networkSuccesses) == 0 {
+		return 1.0
+	}
+
+	errorRate := float64(len(p.networkErrors)) / float64(len(p.networkSuccesses)+len(p.networkErrors))
+	return errorRate
+}
+
+func (p *parallelism) changeParallelism(desired uint) {
+	// Always add the desired to our previous entries.
+	defer func() {
+		p.previous = append(p.previous, previousDesired{
+			desired:  desired,
+			recorded: time.Now(),
+		})
+		p.statshub.SendParralelismStats(types.ParralelismStats{
+			MaxConnections:     p.cfg.MaxConnections,
+			MinConnections:     p.cfg.MinConnections,
+			DesiredConnections: desired,
+		})
+	}()
+	if desired == p.currentDesired {
+		level.Debug(p.l).Log("msg", "desired is equal to current", "desired", desired)
+		return
+	}
+	actualValue := desired
+	// Are we ramping down?
+	if desired < p.currentDesired {
+		// Let's see if we are going to catch the flapping issues.
+		previousInLookback := make([]previousDesired, 0, len(p.previous))
+		for _, previous := range p.previous {
+			// Remove any outliers
+			if time.Since(previous.recorded) <= p.cfg.Lookback {
+				previousInLookback = append(previousInLookback, previous)
+			}
+		}
```

Comment:
This cleanup of `p.previous` I think should be part of `desiredLoops()` so that it doesn't grow indefinitely. Currently if we keep always exiting from this function at `if desired == p.currentDesired` check, we will be growing this array forever.

Reply:
Good point, will move.